src/deps-index.ts

```diff
@@ -21,79 +20,86 @@ export interface DepsIndex {
+    document: Document,
+    viaEager: boolean,
+    visited: Set<UrlString>,
+    visitedEager: Set<UrlString>,
+    lazyImports: Set<UrlString>) {
+  const htmlImports = document.getFeatures(
+      {kind: 'html-import', imported: false, externalPackages: true});
+  for (const htmlImport of htmlImports) {
+    const importUrl = htmlImport.document.url;
+    if (htmlImport.lazy) {
+      lazyImports.add(importUrl);
+    }
+    if (visitedEager.has(importUrl)) {
+      continue;
+    }
+    const isEager = viaEager && !htmlImport.lazy;
+    if (isEager) {
+      visitedEager.add(importUrl);
+      // In this case we've visited a node eagerly for the first time,
+      // so recurse
+    } else if (visited.has(importUrl)) {
+      // In this case we're seeing a node lazily again, so don't recurse
+      continue;
     }
+    visited.add(importUrl);
+    _getHtmlDependencies(
+        htmlImport.document, isEager, visited, visitedEager, lazyImports);
+  }
+}
 
+/**
+ * Analyzes all entrypoints and determines each of their transitive
+ * dependencies.
+ * @param entrypoints Urls of entrypoints to analyze.
+ * @param analyzer
+ * @return a dependency index of every entrypoint, including entrypoints that
+ *     were discovered as lazy entrypoints in the graph.
+ */
 export async function buildDepsIndex(
     entrypoints: UrlString[], analyzer: Analyzer): Promise<DepsIndex> {
-  const entrypointToDependencies: Map<UrlString, Set<UrlString>> = new Map();
-  const dependenciesToEntrypoints: Map<UrlString, Set<UrlString>> = new Map();
-  const queue = Array.from(entrypoints);
-  const visitedEntrypoints = new Set<UrlString>();
-  while (queue.length > 0) {
-    const entrypoint = queue.shift()!;
-    if (visitedEntrypoints.has(entrypoint)) {
-      continue;
+  const depsIndex = {entrypointToDeps: new Map<UrlString, Set<UrlString>>()};
+  const analysis = await analyzer.analyze(entrypoints);
+  const allEntrypoints = new Set<UrlString>(entrypoints);
+  for (const entrypoint of allEntrypoints) {
+    const document = analysis.getDocument(entrypoint);
+    if (!(document instanceof Document)) {
+      const message = document && document.message || '';
+      throw new Error(`Unable to get document ${entrypoint}: ${message}`);
     }
-    const dependencyEntry =
-        await _getTransitiveDependencies(entrypoint, entrypoints, analyzer);
-    const dependencies = new Set(dependencyEntry.eager);
-    dependencies.add(entrypoint);
-    entrypointToDependencies.set(entrypoint, dependencies);
-    for (const lazyDependency of dependencyEntry.lazy.values()) {
-      if (!visitedEntrypoints.has(lazyDependency)) {
-        queue.push(lazyDependency);
-      }
+    const deps = getHtmlDependencies(document);
+
+    depsIndex.entrypointToDeps.set(
+        entrypoint, new Set([entrypoint, ...deps.eagerDeps]));
+
+    // Add lazy imports as entrypoints to index.
+    for (const dep of deps.lazyImports) {
+      allEntrypoints.add(dep);
```

Comment:
does this cause the for loop to iterate over the lazy import? Do we need to process it? I'd consider making this explicit by making a queue of entrypoints and adding the lay import to the queue.